src/empirical.jl

```diff
@@ -55,26 +55,65 @@ struct EmpiricalVariogram
     xi = MVector{ndims(sdata),coordtype(sdata)}(undef)
     xj = MVector{ndims(sdata),coordtype(sdata)}(undef)
 
-    @inbounds (
-    for j in 1:npts
-      coordinates!(xj, sdata, j)
-      for i in j+1:npts
-        coordinates!(xi, sdata, i)
-
-        # evaluate lag and (cross-)variance
-        h = evaluate(distance, xi, xj)
-        v = (z₁[i] - z₁[j])*(z₂[i] - z₂[j])
-
-        # bin (or lag) where to accumulate result
-        lag = ceil(Int, h / Δh)
-
-        if lag ≤ nlags && !ismissing(v) && !isnan(v)
-          sums[lag] += v
-          counts[lag] += 1
+    # decide whether to construct a spatial index
+    # coordinates need to be AbstractFloat for NearestNeighbors to work
+    # https://github.com/KristofferC/NearestNeighbors.jl/issues/13
+    isfloat = coordtype(sdata) <: AbstractFloat
+    # quite arbitrary, but for small datasets setting up the index is too much overhead
+    # it really also depends on what fraction of points are within hmax
+    isbig = npts > 1000
+
+    if isfloat && isbig
+      # construct the variogram with neighbour search
```

Comment:
What about creating an additional function called `spatial_index_variogram` for this branch, and another function `simple_loop_variogram`? These could then be more easily tested in the test suite.